share/spack/gitlab/cloud_pipelines/configs/config.yaml

```diff
@@ -1,5 +1,8 @@
 config:
   db_lock_timeout: 120
+  # Ref. share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml::.generate-common
+  # 2x over subscription on request for generate job concretization CPU pool
+  build_jobs: 8
```

Comment:
How does this prevent an over subscription? Isn't it just reducing the number of procs per generate job?

Reply:
For the record the originating issue is that these two jobs:
- https://gitlab.spack.io/spack/spack/-/jobs/13721155
- https://gitlab.spack.io/spack/spack/-/jobs/13721154

were running concurrently on the same node.

Reply:
This sets the max number of concretization jobs that can run. We set the allocation request to 4 vCPUs and 16GB memory. This says use up to 8 jobs (ie. 8 CPUs) during concretize.

As the comment implies, the permits a "2x over subscription" based on the request set in the generate job. So to answer

> How does this prevent an over subscription?

I am guessing this is asking about assigning multiple jobs to the same kube node.

It does not 😃 We allow multiple build/generate jobs to run on the same node if the requests allow it. This is important as we want to pack jobs onto allocated resources as much as possible. This is the point of setting the `KUBERNETES_MEMORY_REQUEST` and `KUBERNETES_CPU_REQUEST` on jobs. We say "I promise I will use this much", and then pack everything we can on there and set the `build_jobs` to that request in the case of build jobs.

> Isn't it just reducing the number of procs per generate job?

Yes. The default was setting the CPU allocation to the min of 16 or the number of CPUs available on the node. What we want instead is to set that to a number that is aware of the allocation we are requesting. In the case of generate, we are requesting 4 vCPUs. My proposal is to permit up to **8 concretize tasks** to run concurrently. I pick this number because 2x is usually a pretty safe over subscription number as CPUs generally handling that pretty well.

I am also happy to drop it if we think that that is too high.

Reply:
I was not aware the allocation request on the infra side was for 4 vCPUs. If we could restrict the processes to that, then this:

https://github.com/spack/spack/blob/dfab174f3100840c889e8bb939260b64d93d8dbd/lib/spack/spack/util/cpus.py#L10-L20

should already take care of the maximum, right?

Reply:
> If we could restrict the processes to that, then this should already take care of the maximum, right?

It does not in this case. We set a "requests" on the nodes, but not a "requirements". The reason for this is we don't want to kill jobs if they exceed these limits, we just try to configure Spack to operate with in them as best we can.

Reply:
> The reason for this is we don't want to kill jobs if they exceed these limits, we just try to configure Spack to operate with in them as best we can.

I understand why this makes sense for memory, but isn't there anything like `taskset` to set the CPU affinity for the processes spawned in a given job?

Reply:
somewhat related, I'm starting to play around with setting CPU limits for build jobs, which would theoretically prevent jobs from using resources "promised" to others on the same node not by killing, but throttling.

> The CPU limit defines a hard ceiling on how much CPU time the container can use. During each scheduling interval (time slice), the Linux kernel checks to see if this limit is exceeded; if so, the kernel waits before allowing that cgroup to resume execution.


https://github.com/spack/spack-infrastructure/pull/999 enables this functionality